### PR TITLE
Added route for deleting slide from file system

### PR DESCRIPTION
**Pull request by Vedant1202**
This update contains:

1. Deletion route for slide from system files.

This is a fix for [#324](https://github.com/camicroscope/caMicroscope/issues/324) on caMicrscope repository.

### CODE REVIEW

**Comment by birm** [comment]
I'm a little nervous about exposing file deletion functionality. Maybe if we limited to admin user types?

**Comment by Vedant1202** [comment]
@birm Yes, we could do that, but then should it be implemented on the frontend as well? Like the delete button only gets shown to admin users or else it becomes messy.

**Comment by Vedant1202** [comment]
Or else, we could implement a review system, just like the PR system in github? A regular user can request the admin to delete a slide and the admin can approve it

**Comment by cjchirag7** [comment]
I want to point out some issues and vulnerabilities with this PR :
- This PR doesn't perform any check whether the user, sending the request is admin or not. Even there is no check to find if the user is known or any anonymous user. This may result into serious security issues in future.
- One more thing, it allows the request to be performed even on 'GET' request. This is very dangerous. Suppose, someone types the URL for '/slide/delete' by mistake, or somehow finds out that this is the delete endpoint. Then, he/she may be able to delete all the slides from the file system, without any permissions.
- If you disallow 'GET' request on this endpoint, even then, this is not secure, as someone might easily use the browser dev tools to get the code for delete (the endpoint and the request format), and may send a 'POST' request via curl, postman, etc. to delete all the slides.

**Comment by cjchirag7** [comment]
> @birm Yes, we could do that, but then should it be implemented on the frontend as well? Like the delete button only gets shown to admin users or else it becomes messy.

I would like to inform you that , we are already checking it in the frontend and displaying the delete button, accordingly in the frontend. However, it should be taken care that the frontend code is visible to everyone easily, via the browser dev tools and as mentioned earlier, this might result into serious issues, if no checking for admin is performed at the backend.

**Comment by cjchirag7** [comment]
Considering this issues, I think [#23](https://github.com/camicroscope/Caracal/pull/23) is a better fix, as it takes care that when the slide is deleted from the DB, then only the file is deleted locally, and keeps the route secure as earlier.

**Comment by Vedant1202** [comment]
This update contains:

1. Security check functions for SlideLoader.
2. Secured Delete route, only available to admins.

**Comment by Vedant1202** [comment]
@cjchirag7 I've added the security functions so it solves the issue I guess. 
Also keeping slide related system functions in one place seems to be a better idea than putting them in another repository.

**Comment by Vedant1202** [comment]
This feature depends on [PR-119](https://github.com/camicroscope/Distro/pull/119)

**Comment by birm** [comment]
Both of these techniques end up introducing a function from one service into the other. Couldn't we add the permission handler to the delete route of the loader service specifically?

**Comment by Vedant1202** [comment]
@birm I may have not caught upon exactly what you were trying to say there, but the permission handler is serving as a wrapper around the delete route. It does'nt call upon another service to authenticate. 
If you are talking about integrating authentication inside the route, then the route still remains exposed to DoS attacks as the any invalid request will still get to the route. A wrapper dissolves that issue.
Also, it's modifiable and can be used similarly for other services in the loader.

**Comment by birm** [comment]
I agree; what I mean is that we have the backend (caracal) which is acting as a proxy to this service; we can deny the delete route to unauthorized parties there by adding a handler specific to this action in caracal. That saves us from having to do any token checking here.

**Comment by Vedant1202** [comment]
@birm Do you mean [adding such a line](https://github.com/camicroscope/Caracal/blob/837ea0a7fa244d14b507200dd54dfe9f86da9e0a/caracal.js#L52) and specifying only "Admin" to the delete route?

**Comment by cjchirag7** [comment]
@birm , the token checking is already being done for all /loader/ routes [here](https://github.com/camicroscope/Caracal/blob/d80ea8198f782441d5017ef21ac80ce2f59de88f/caracal.js#L52) . So, we can directly send the request, along with the token. And it is already being checked for 'Editor' and 'Admin'. So, we don't need to make any change there, I suppose (since both Editor and Admin are allowed to delete a slide)

**Comment by Vedant1202** [comment]
@cjchirag7 Yes, the checking has been done but it still poses a threat when the route is hit outside of the client which uses the proxy url.

**Comment by Vedant1202** [comment]
@birm I had thought of adding checker to caracal before as well, but the ip stays unprotected. 

Screencast-from-Thursday-09-April-2020-02_04_10-IST

Like this example where the unprotected route for deleting slide on slideloader is being hit outside of the client by using it's ip instead of the proxy url. The slide 'sample.svs' gets deleted even though no tokenisation is perfomed

Such attacks are common on proxies and thus each individual server having their own token checking mechanism proves is better for additional security.

_Note that unprotected here means the token checker which I implemented for slideloader has been removed from that route. Other security measures in caracal are in place and have been configured to be active_

**Comment by Vedant1202** [comment]
I've mentioned this in [#27 ](https://github.com/camicroscope/SlideLoader/issues/27). Please have a look.

**Comment by Vedant1202** [comment]
@birm Roger that, I'm working on it and thought I will push all the changes together. 
Just keep the PR live, I'll report back to you when I'm done.

**Comment by Vedant1202** [comment]
@birm I've done the changes you can review them.

File: SlideServer.py [hunk 1]
```diff
@@ -6,13 +6,24 @@
...
 import dev_utils
 
+from flask import Flask, request, jsonify, _request_ctx_stack
+from flask_cors import cross_origin
+
+AUTH0_DOMAIN = 'YOUR_DOMAIN'
+API_AUDIENCE = False
+ALGORITHMS = ["RS256"]
+
+
```
Comment (birm) [inline SlideServer.py:25]:
```suggestion
```
I don't think we're using any of this, right?

File: SlideServer.py [hunk 2]
```diff
@@ -130,6 +141,29 @@ def finish_upload(token):
...
+        
+    filename = body['filename']
+    if filename and allowed_file(filename):
+        filename = secure_filename(filename)
+        filepath = os.path.join(app.config['UPLOAD_FOLDER'], filename)
+        if os.path.isfile(filepath):
+            os.remove(os.path.join(app.config['UPLOAD_FOLDER'], filename))
+            return flask.Response(json.dumps({"deleted": filename, "success": True})) 
+        else:
+            return flask.Response(json.dumps({"error": "File with name '" + filename + "' does not exists"}), status=400)
```
Comment (birm) [inline SlideServer.py:160]:
```suggestion
            return flask.Response(json.dumps({"error": "File with name '" + filename + "' does not exist"}), status=400)
```

File: SlideServer.py (hunk 1 above)
Comment (Vedant1202) [reply SlideServer.py:25]: Yeah, sorry missed it

File: SlideServer.py (hunk 2 above)
Comment (Vedant1202) [reply SlideServer.py:152]: Oh, grammar :/

**Pull request merged**